File: shared/src/main/java/com/runtracker/shared/data/model/MentalHealth.kt
package com.runtracker.shared.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * Mood entry for tracking emotional state before/after workouts
 */
@Entity(tableName = "mood_entries")
data class MoodEntry(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val timestamp: Long = System.currentTimeMillis(),
    val mood: MoodLevel,
    val energy: EnergyLevel,
    val stress: StressLevel,
    val notes: String = "",
    val relatedRunId: Long? = null,
    val isPreWorkout: Boolean = true,
    val tags: List<String> = emptyList()
)

enum class MoodLevel(val value: Int, val emoji: String, val label: String) {
    VERY_LOW(1, "😢", "Very Low"),
    LOW(2, "😔", "Low"),
    NEUTRAL(3, "😐", "Neutral"),
    GOOD(4, "🙂", "Good"),
    GREAT(5, "😄", "Great")
}

enum class EnergyLevel(val value: Int, val emoji: String, val label: String) {
    EXHAUSTED(1, "😴", "Exhausted"),
    TIRED(2, "🥱", "Tired"),
    MODERATE(3, "😊", "Moderate"),
    ENERGIZED(4, "💪", "Energized"),
    PUMPED(5, "🔥", "Pumped")
}

enum class StressLevel(val value: Int, val emoji: String, val label: String) {
    VERY_HIGH(5, "😰", "Very High"),
    HIGH(4, "😟", "High"),
    MODERATE(3, "😐", "Moderate"),
    LOW(2, "😌", "Low"),
    VERY_LOW(1, "😊", "Very Low")
}

/**
 * Mindfulness/meditation session record
 */
@Entity(tableName = "mindfulness_sessions")
data class MindfulnessSession(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val timestamp: Long = System.currentTimeMillis(),
    val type: MindfulnessType,
    val durationSeconds: Int,
    val completed: Boolean = true,
    val relatedRunId: Long? = null,
    val rating: Int? = null // 1-5 stars
)

enum class MindfulnessType(val title: String, val description: String, val defaultDuration: Int) {
    PRE_RUN_FOCUS("Pre-Run Focus", "Clear your mind and set intentions for your run", 180),
    POST_RUN_GRATITUDE("Post-Run Gratitude", "Reflect on your accomplishment and express gratitude", 180),
    BREATHING_EXERCISE("Breathing Exercise", "Calm your nervous system with guided breathing", 120),
    BODY_SCAN("Body Scan", "Check in with your body from head to toe", 300),
    VISUALIZATION("Visualization", "Visualize your perfect run or race", 240),
    STRESS_RELIEF("Stress Relief", "Release tension and find calm", 300),
    SLEEP_PREPARATION("Sleep Preparation", "Wind down for restful recovery sleep", 600),
    MORNING_ENERGIZER("Morning Energizer", "Start your day with positive energy", 180)
}

/**
 * Breathing exercise pattern
 */
data class BreathingPattern(
    val name: String,
    val description: String,
    val inhaleSeconds: Int,
    val holdAfterInhale: Int,
    val exhaleSeconds: Int,
    val holdAfterExhale: Int,
    val cycles: Int,
    val benefits: List<String>
) {
    val cycleDuration: Int
        get() = inhaleSeconds + holdAfterInhale + exhaleSeconds + holdAfterExhale
    
    val totalDuration: Int
        get() = cycleDuration * cycles
}

/**
 * Pre-defined breathing patterns
 */
object BreathingPatterns {
    val boxBreathing = BreathingPattern(
        name = "Box Breathing",
        description = "Equal parts inhale, hold, exhale, hold. Used by Navy SEALs for stress relief.",
        inhaleSeconds = 4,
        holdAfterInhale = 4,
        exhaleSeconds = 4,
        holdAfterExhale = 4,
        cycles = 6,
        benefits = listOf("Reduces stress", "Improves focus", "Calms nervous system")
    )
    
    val relaxingBreath = BreathingPattern(
        name = "4-7-8 Relaxing Breath",
        description = "Dr. Andrew Weil's technique for relaxation and sleep.",
        inhaleSeconds = 4,
        holdAfterInhale = 7,
        exhaleSeconds = 8,
        holdAfterExhale = 0,
        cycles = 4,
        benefits = listOf("Promotes sleep", "Reduces anxiety", "Lowers heart rate")
    )
    
    val energizingBreath = BreathingPattern(
        name = "Energizing Breath",
        description = "Quick, powerful breaths to increase alertness before a run.",
        inhaleSeconds = 2,
        holdAfterInhale = 0,
        exhaleSeconds = 2,
        holdAfterExhale = 0,
        cycles = 15,
        benefits = listOf("Increases energy", "Improves alertness", "Warms up lungs")
    )
    
    val calmingBreath = BreathingPattern(
        name = "Extended Exhale",
        description = "Longer exhale activates parasympathetic nervous system.",
        inhaleSeconds = 4,
        holdAfterInhale = 0,
        exhaleSeconds = 8,
        holdAfterExhale = 2,
        cycles = 6,
        benefits = listOf("Activates rest response", "Slows heart rate", "Reduces cortisol")
    )
    
    val recoveryBreath = BreathingPattern(
        name = "Post-Run Recovery",
        description = "Gentle breathing to aid recovery after intense exercise.",
        inhaleSeconds = 5,
        holdAfterInhale = 2,
        exhaleSeconds = 5,
        holdAfterExhale = 2,
        cycles = 8,
        benefits = listOf("Speeds recovery", "Reduces lactic acid", "Restores calm")
    )
    
    fun getAll(): List<BreathingPattern> = listOf(
        boxBreathing, relaxingBreath, energizingBreath, calmingBreath, recoveryBreath
    )
    
    fun getPreRunPatterns(): List<BreathingPattern> = listOf(energizingBreath, boxBreathing)
    fun getPostRunPatterns(): List<BreathingPattern> = listOf(recoveryBreath, calmingBreath)
    fun getSleepPatterns(): List<BreathingPattern> = listOf(relaxingBreath, calmingBreath)
}

/**
 * Daily wellness check-in
 */
@Entity(tableName = "wellness_checkins")
data class WellnessCheckin(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val date: Long, // Start of day timestamp
    val sleepHours: Float? = null,
    val sleepQuality: Int? = null, // 1-5
    val restingHeartRate: Int? = null,
    val hrv: Int? = null, // Heart Rate Variability
    val mood: MoodLevel? = null,
    val energy: EnergyLevel? = null,
    val stress: StressLevel? = null,
    val soreness: Int? = null, // 1-5, muscle soreness
    val hydration: Int? = null, // 1-5
    val notes: String = "",
    val readinessScore: Int? = null // Calculated 0-100
)

/**
 * Guided meditation/mindfulness content
 */
data class MindfulnessContent(
    val id: String,
    val title: String,
    val description: String,
    val type: MindfulnessType,
    val durationSeconds: Int,
    val audioUrl: String? = null,
    val instructions: List<String>,
    val tags: List<String> = emptyList()
)

/**
 * Pre-defined mindfulness sessions
 */
object MindfulnessSessions {
    
    val preRunFocus = MindfulnessContent(
        id = "pre_run_focus",
        title = "Pre-Run Mental Preparation",
        description = "Get mentally ready for your run with focused breathing and intention setting.",
        type = MindfulnessType.PRE_RUN_FOCUS,
        durationSeconds = 180,
        instructions = listOf(
            "Find a comfortable standing or seated position",
            "Close your eyes and take 3 deep breaths",
            "Set your intention for today's run",
            "Visualize yourself running strong and confident",
            "Feel gratitude for your body's ability to move",
            "Open your eyes, ready to run"
        ),
        tags = listOf("pre-run", "focus", "intention")
    )
    
    val postRunGratitude = MindfulnessContent(
        id = "post_run_gratitude",
        title = "Post-Run Reflection",
        description = "Celebrate your accomplishment and aid recovery with mindful reflection.",
        type = MindfulnessType.POST_RUN_GRATITUDE,
        durationSeconds = 180,
        instructions = listOf(
            "Find a comfortable position to cool down",
            "Close your eyes and notice your breath returning to normal",
            "Acknowledge what you just accomplished",
            "Thank your body for carrying you through",
            "Notice how you feel - physically and emotionally",
            "Set an intention for recovery"
        ),
        tags = listOf("post-run", "gratitude", "recovery")
    )
    
    val bodyScan = MindfulnessContent(
        id = "body_scan",
        title = "Runner's Body Scan",
        description = "Check in with your body to identify tension and promote recovery.",
        type = MindfulnessType.BODY_SCAN,
        durationSeconds = 300,
        instructions = listOf(
            "Lie down or sit comfortably",
            "Start at the top of your head, notice any tension",
            "Move attention to your face, jaw, neck",
            "Scan your shoulders, arms, hands",
            "Notice your chest, breathing naturally",
            "Feel your core, lower back",
            "Scan your hips, glutes",
            "Move to your thighs, knees",
            "Notice your calves, ankles, feet",
            "Take a moment to feel your whole body as one"
        ),
        tags = listOf("body-scan", "recovery", "awareness")
    )
    
    val stressRelief = MindfulnessContent(
        id = "stress_relief",
        title = "Quick Stress Relief",
        description = "Release tension and find calm in just 5 minutes.",
        type = MindfulnessType.STRESS_RELIEF,
        durationSeconds = 300,
        instructions = listOf(
            "Sit or stand comfortably",
            "Take 3 deep breaths, exhaling fully",
            "Tense your shoulders up to your ears, hold, release",
            "Clench your fists tight, hold, release",
            "Scrunch your face, hold, release",
            "Feel the wave of relaxation through your body",
            "Continue breathing slowly and deeply",
            "Imagine stress leaving your body with each exhale"
        ),
        tags = listOf("stress", "relaxation", "quick")
    )
    
    fun getAll(): List<MindfulnessContent> = listOf(
        preRunFocus, postRunGratitude, bodyScan, stressRelief
    )
    
    fun getPreRunSessions(): List<MindfulnessContent> = listOf(preRunFocus)
    fun getPostRunSessions(): List<MindfulnessContent> = listOf(postRunGratitude, bodyScan)
}

File: iosApp/GoSteady/Sources/Models/MentalHealth.swift
import Foundation
import SwiftData

@Model
final class MoodEntry {
    @Attribute(.unique) var id: UUID
    var timestamp: Date
    var mood: MoodLevel
    var energy: EnergyLevel
    var stress: StressLevel
    var notes: String
    var relatedRunId: UUID?
    var isPreWorkout: Bool
    var tags: [String]

    init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        mood: MoodLevel,
        energy: EnergyLevel,
        stress: StressLevel,
        notes: String = "",
        relatedRunId: UUID? = nil,
        isPreWorkout: Bool = true,
        tags: [String] = []
    ) {
        self.id = id
        self.timestamp = timestamp
        self.mood = mood
        self.energy = energy
        self.stress = stress
        self.notes = notes
        self.relatedRunId = relatedRunId
        self.isPreWorkout = isPreWorkout
        self.tags = tags
    }
}

enum MoodLevel: Int, Codable, CaseIterable {
    case veryLow = 1
    case low = 2
    case neutral = 3
    case good = 4
    case great = 5

    var emoji: String {
        switch self {
        case .veryLow: return "\u{1F622}"
        case .low: return "\u{1F614}"
        case .neutral: return "\u{1F610}"
        case .good: return "\u{1F642}"
        case .great: return "\u{1F604}"
        }
    }

    var label: String {
        switch self {
        case .veryLow: return "Very Low"
        case .low: return "Low"
        case .neutral: return "Neutral"
        case .good: return "Good"
        case .great: return "Great"
        }
    }
}

enum EnergyLevel: Int, Codable, CaseIterable {
    case exhausted = 1
    case tired = 2
    case moderate = 3
    case energized = 4
    case pumped = 5

    var emoji: String {
        switch self {
        case .exhausted: return "\u{1F634}"
        case .tired: return "\u{1F971}"
        case .moderate: return "\u{1F60A}"
        case .energized: return "\u{1F4AA}"
        case .pumped: return "\u{1F525}"
        }
    }

    var label: String {
        switch self {
        case .exhausted: return "Exhausted"
        case .tired: return "Tired"
        case .moderate: return "Moderate"
        case .energized: return "Energized"
        case .pumped: return "Pumped"
        }
    }
}

enum StressLevel: Int, Codable, CaseIterable {
    case veryHigh = 5
    case high = 4
    case moderate = 3
    case low = 2
    case veryLow = 1

    var emoji: String {
        switch self {
        case .veryHigh: return "\u{1F630}"
        case .high: return "\u{1F61F}"
        case .moderate: return "\u{1F610}"
        case .low: return "\u{1F60C}"
        case .veryLow: return "\u{1F60A}"
        }
    }

    var label: String {
        switch self {
        case .veryHigh: return "Very High"
        case .high: return "High"
        case .moderate: return "Moderate"
        case .low: return "Low"
        case .veryLow: return "Very Low"
        }
    }
}

@Model
final class MindfulnessSession {
    @Attribute(.unique) var id: UUID
    var timestamp: Date
    var type: MindfulnessType
    var durationSeconds: Int
    var completed: Bool
    var relatedRunId: UUID?
    var rating: Int?

    init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        type: MindfulnessType,
        durationSeconds: Int,
        completed: Bool = true,
        relatedRunId: UUID? = nil,
        rating: Int? = nil
    ) {
        self.id = id
        self.timestamp = timestamp
        self.type = type
        self.durationSeconds = durationSeconds
        self.completed = completed
        self.relatedRunId = relatedRunId
        self.rating = rating
    }
}

enum MindfulnessType: String, Codable, CaseIterable {
    case preRunFocus = "PRE_RUN_FOCUS"
    case postRunGratitude = "POST_RUN_GRATITUDE"
    case breathingExercise = "BREATHING_EXERCISE"
    case bodyScan = "BODY_SCAN"
    case visualization = "VISUALIZATION"
    case stressRelief = "STRESS_RELIEF"
    case sleepPreparation = "SLEEP_PREPARATION"
    case morningEnergizer = "MORNING_ENERGIZER"

    var title: String {
        switch self {
        case .preRunFocus: return "Pre-Run Focus"
        case .postRunGratitude: return "Post-Run Gratitude"
        case .breathingExercise: return "Breathing Exercise"
        case .bodyScan: return "Body Scan"
        case .visualization: return "Visualization"
        case .stressRelief: return "Stress Relief"
        case .sleepPreparation: return "Sleep Preparation"
        case .morningEnergizer: return "Morning Energizer"
        }
    }

    var typeDescription: String {
        switch self {
        case .preRunFocus: return "Clear your mind and set intentions for your run"
        case .postRunGratitude: return "Reflect on your accomplishment and express gratitude"
        case .breathingExercise: return "Calm your nervous system with guided breathing"
        case .bodyScan: return "Check in with your body from head to toe"
        case .visualization: return "Visualize your perfect run or race"
        case .stressRelief: return "Release tension and find calm"
        case .sleepPreparation: return "Wind down for restful recovery sleep"
        case .morningEnergizer: return "Start your day with positive energy"
        }
    }

    var defaultDuration: Int {
        switch self {
        case .preRunFocus: return 180
        case .postRunGratitude: return 180
        case .breathingExercise: return 120
        case .bodyScan: return 300
        case .visualization: return 240
        case .stressRelief: return 300
        case .sleepPreparation: return 600
        case .morningEnergizer: return 180
        }
    }

    var icon: String {
        switch self {
        case .preRunFocus: return "figure.run"
        case .postRunGratitude: return "heart.fill"
        case .breathingExercise: return "wind"
        case .bodyScan: return "person.fill"
        case .visualization: return "eye.fill"
        case .stressRelief: return "leaf.fill"
        case .sleepPreparation: return "moon.fill"
        case .morningEnergizer: return "sun.max.fill"
        }
    }
}

struct BreathingPattern: Codable, Hashable, Identifiable {
    var id: String { name }
    var name: String
    var patternDescription: String
    var inhaleSeconds: Int
    var holdAfterInhale: Int
    var exhaleSeconds: Int
    var holdAfterExhale: Int
    var cycles: Int
    var benefits: [String]

    var cycleDuration: Int {
        inhaleSeconds + holdAfterInhale + exhaleSeconds + holdAfterExhale
    }

    var totalDuration: Int {
        cycleDuration * cycles
    }
}

enum BreathingPatterns {
    static let boxBreathing = BreathingPattern(
        name: "Box Breathing",
        patternDescription: "Equal parts inhale, hold, exhale, hold. Used by Navy SEALs for stress relief.",
        inhaleSeconds: 4, holdAfterInhale: 4, exhaleSeconds: 4, holdAfterExhale: 4, cycles: 6,
        benefits: ["Reduces stress", "Improves focus", "Calms nervous system"]
    )

    static let relaxingBreath = BreathingPattern(
        name: "4-7-8 Relaxing Breath",
        patternDescription: "Dr. Andrew Weil's technique for relaxation and sleep.",
        inhaleSeconds: 4, holdAfterInhale: 7, exhaleSeconds: 8, holdAfterExhale: 0, cycles: 4,
        benefits: ["Promotes sleep", "Reduces anxiety", "Lowers heart rate"]
    )

    static let energizingBreath = BreathingPattern(
        name: "Energizing Breath",
        patternDescription: "Quick, powerful breaths to increase alertness before a run.",
        inhaleSeconds: 2, holdAfterInhale: 0, exhaleSeconds: 2, holdAfterExhale: 0, cycles: 15,
        benefits: ["Increases energy", "Improves alertness", "Warms up lungs"]
    )

    static let calmingBreath = BreathingPattern(
        name: "Extended Exhale",
        patternDescription: "Longer exhale activates parasympathetic nervous system.",
        inhaleSeconds: 4, holdAfterInhale: 0, exhaleSeconds: 8, holdAfterExhale: 2, cycles: 6,
        benefits: ["Activates rest response", "Slows heart rate", "Reduces cortisol"]
    )

    static let recoveryBreath = BreathingPattern(
        name: "Post-Run Recovery",
        patternDescription: "Gentle breathing to aid recovery after intense exercise.",
        inhaleSeconds: 5, holdAfterInhale: 2, exhaleSeconds: 5, holdAfterExhale: 2, cycles: 8,
        benefits: ["Speeds recovery", "Reduces lactic acid", "Restores calm"]
    )

    static func getAll() -> [BreathingPattern] {
        [boxBreathing, relaxingBreath, energizingBreath, calmingBreath, recoveryBreath]
    }

    static func getPreRunPatterns() -> [BreathingPattern] { [energizingBreath, boxBreathing] }
    static func getPostRunPatterns() -> [BreathingPattern] { [recoveryBreath, calmingBreath] }
    static func getSleepPatterns() -> [BreathingPattern] { [relaxingBreath, calmingBreath] }
}

@Model
final class WellnessCheckin {
    @Attribute(.unique) var id: UUID
    var date: Date
    var sleepHours: Float?
    var sleepQuality: Int?
    var restingHeartRate: Int?
    var hrv: Int?
    var mood: MoodLevel?
    var energy: EnergyLevel?
    var stress: StressLevel?
    var soreness: Int?
    var hydration: Int?
    var notes: String
    var readinessScore: Int?

    init(
        id: UUID = UUID(),
        date: Date,
        sleepHours: Float? = nil,
        sleepQuality: Int? = nil,
        restingHeartRate: Int? = nil,
        hrv: Int? = nil,
        mood: MoodLevel? = nil,
        energy: EnergyLevel? = nil,
        stress: StressLevel? = nil,
        soreness: Int? = nil,
        hydration: Int? = nil,
        notes: String = "",
        readinessScore: Int? = nil
    ) {
        self.id = id
        self.date = date
        self.sleepHours = sleepHours
        self.sleepQuality = sleepQuality
        self.restingHeartRate = restingHeartRate
        self.hrv = hrv
        self.mood = mood
        self.energy = energy
        self.stress = stress
        self.soreness = soreness
        self.hydration = hydration
        self.notes = notes
        self.readinessScore = readinessScore
    }
}

struct MindfulnessContent: Codable, Hashable, Identifiable {
    var id: String
    var title: String
    var contentDescription: String
    var type: MindfulnessType
    var durationSeconds: Int
    var audioUrl: String?
    var instructions: [String]
    var tags: [String]

    init(
        id: String,
        title: String,
        contentDescription: String,
        type: MindfulnessType,
        durationSeconds: Int,
        audioUrl: String? = nil,
        instructions: [String],
        tags: [String] = []
    ) {
        self.id = id
        self.title = title
        self.contentDescription = contentDescription
        self.type = type
        self.durationSeconds = durationSeconds
        self.audioUrl = audioUrl
        self.instructions = instructions
        self.tags = tags
    }
}
